keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1899 @@
+  management of QoS resources is handled in the container runtime
+  - no "system reserved" (or equivalent) exists (considered as configuration
+    detail outside Kubernetes)
+  - no overprovisioning (or auto-promotion to free classes) exists (considered
+    as implementation detail of specific QoS-class resource, outside
+    Kubernetes)
+  - not requesting anything implies a "default" that is an implementation
+    detail on the container runtime side
+- Pod and container level QoS-class resources are separate concepts
+  - Pod-level QoS-class resources are meant for sandbox-level QoS
+  - Pod-level QoS-class resources cannot be specified in container resources field
+  - container-level QoS-class resources are meant for per-container QoS
+  - container-level QoS-class resources can be specified in pod-level request
+    field (in PodSpec) in which case it is regarded as a default for all
+    containers within that pod
+  - resource names across pod and container level QoS-class resource names
+    must be unique - i.e. the runtime cannot register a pod-level and
+    container-level resource with the same name
+- runtime advertises available QoS resources to kubelet
+  - QoS resources are dynamic i.e. can change during the lifetime of the node
+  - CRI [RuntimeStatus](#runtimestatus) message is used for carrying the information
+- kubelet
+  - updates NodeStatus based on the available QoS-class resources advertised by
+    the runtimea
+  - relays QoS-class resource requests from PodSpec to the runtime at pod
+    startup via [ContainerConfig](#containerconfig) and
+    [PodSandboxConfig](#podsandboxconfig) messages
+  - implements admission handler that validates the availability of QoS-class resources
+  - pod eviction as a possible [future improvement](#kubelet-initiated-pod-eviction)
+- scheduler
+  - simple node filtering based on the availability of QoS-class resources in
+    the first implementation phase (with possible
+    [future improvements](#scheduler-improvements))
+  - pod priority based eviction is in effect
+- quota, limits and defaults are [future work](#future-work) items
+
+
+### CRI API
+
+The following additions to the CRI protocol are suggested.
+
+#### ContainerConfig
+
+The `ContainerConfig` message will be supplemented with new `class_resources`
+field, providing per-container setting for QoS-class resources. This will be
+used in `CreateContainerRequest` to communicate the container-level QoS-class
+resource assignments to the runtime.
+
+```diff
+ message ContainerConfig {
+
+ ...
+     // Configuration specific to Linux containers.
+     LinuxContainerConfig linux = 15;
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
++
++    // Configuration of QoS resources.
++    ContainerQoSResources qos_resources = 17;
+ }
+
++// ContainerQoSResources specifies the configuration of QoS resources of a
++// container.
++message ContainerQoSResources {
```

Comment:
I can't imagine containe rand pod QoSResources would be all that different, should they be made the same struct?

Reply:
You mean re-use same message type in two places (pod cri calls and container cri calls) ?

Reply:
Good question, I played safe here (and this separation also reflects the current proposal of K8s API) 😇 You might know better than me, how likely it is that this would bit us in the future. Any thoughts @mrunalp @mikebrow or others?

Reply:
since the types are opaque to kube, I don't see there being any issue in having them be the same. The K/V pairs can be interpreted appropriately by runtime, and presumably contain all the required metadata that would differentiate the two

Reply:
That is true. The only reason for having a separate struct was to keep the door open to differentiate the two in the future in terms of some extra fields (that the other doesn't need/support). I can surely change this to a single type if others agree

Reply:
marked as UNRESOLVED in the latest update

Reply:
-> marking this particular comment thread as resolved